### PR TITLE
Fix naked algorithm

### PR DESCRIPTION
**Pull request by travisdowns**
The naked algorithm doesn't work properly and gives nonsense results (impossibly fast), because we were just populating the indexes with `rand() % length`. When you pointer chase over these indexes, you are likely to quickly settle into a cycle much shorter than the length of the array.

Instead, we now shuffle the array so that it has a cycle of maximum length: every position will be visited before we return to the start.

### CODE REVIEW

File: 2018/11/05/Makefile [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
-	c++ -O2 -o testingmlp testingmlp.cpp
+CXXFLAGS = -O2 -std=c++11
 
+testingmlp: testingmlp.o generated.o
+	$(CXX) $(CXXFLAGS) -o $@ $^
 
 
-test: testingmlp.cpp
+ifneq ($(GENERATE), 0)
+generated.cpp : gen.py
```
Comment (travisdowns) [inline 2018/11/05/Makefile]: The idea here is to generate the `generated.cpp` file in during the build, so you dont' have to manually fix up any code. The `generated.cpp` file is still checked in to git, however, some user doesn't have `python3` they can just run `make GENERATE=0` and use the existing file (of course they can't generate but if they just want to build that's fine).

File: 2018/11/05/common.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <time.h>
+
+/* max MLP value tested with the naked strategy */
+constexpr int NAKED_MAX = 30;
+
+/* if true, try the pure pointer chasing strategy */
+constexpr int DO_NAKED = true;
```
Comment (travisdowns) [inline 2018/11/05/common.hpp]: set this to false for the original test

File: 2018/11/05/testingmlp.cpp [hunk 3]
```diff
@@ -1,1688 +1,33 @@
...
-  uint64_t val28 = 28;
-  uint64_t val29 = 29;
-  size_t howmanyhits_perlane = howmanyhits / 29;
+template <size_t howmanylanes>
+uint64_t access(uint64_t *bigarray, size_t length, size_t howmanyhits) {
+  size_t howmanyhits_perlane = howmanyhits / howmanylanes;
+  uint64_t lanesums[howmanylanes];
+  for (size_t i = 0; i < howmanylanes; i++)
+    lanesums[i] = 0;
   for (size_t counter = 0; counter < howmanyhits_perlane; counter++) {
```
Comment (travisdowns) [inline 2018/11/05/testingmlp.cpp]: I didn't really change any code in this file, except what was needed to extract the generated code out - this is just github matching up the diff weirdly.

**Comment by lemire** [comment]
Great.

**Pull request merged**